bsp/raspberrypi/rp2040/src/hal/spi.zig

```diff
@@ -12,57 +12,141 @@ const hw = @import("hw.zig");
 
 const SpiRegs = microzig.chip.types.peripherals.SPI0;
 
+pub const Motorola = struct {
+    pub const Polarity = enum(u1) {
+        low = 0,
+        high = 1,
+    };
+
+    pub const Phase = enum(u1) {
+        first_edge = 0,
+        second_edge = 1,
+    };
+
+    /// Low means clock is steady state low, high means steady state high
+    clock_polarity: Polarity = .low,
+
+    /// Controls whether data is captured on the first or second clock edge transition
+    clock_phase: Phase = .first_edge,
+};
+
+pub const FrameFormat = union(enum) {
+    motorola: Motorola,
+    texas_instruments,
+    ns_microwire,
+};
+
 pub const Config = struct {
     clock_config: clocks.GlobalConfiguration,
-    tx_pin: ?gpio.Pin = gpio.num(19),
-    rx_pin: ?gpio.Pin = gpio.num(16),
-    sck_pin: ?gpio.Pin = gpio.num(18),
-    csn_pin: ?gpio.Pin = gpio.num(17),
-    baud_rate: u32 = 1000 * 1000,
+    /// Number of bits in a frame can't be less than 4 or more than 16
+    data_width: u4 = 8,
+    frame_format: FrameFormat = .{ .motorola = .{} },
+    baud_rate: u32 = 1_000_000,
 };
 
-pub fn num(n: u1) SPI {
-    return @as(SPI, @enumFromInt(n));
-}
+pub const instance = struct {
+    pub const SPI0: SPI = @as(SPI, @enumFromInt(0));
+    pub const SPI1: SPI = @as(SPI, @enumFromInt(1));
+    pub fn num(instance_number: u1) SPI {
+        return @as(SPI, @enumFromInt(instance_number));
+    }
+};
 
+pub const ConfigError = error{
+    InvalidDataWidth,
+    InputFreqTooLow,
+    InputFreqTooHigh,
+    UnsupportedBaudRate,
+};
+
+/// An API for interacting with the RP2040's SPI driver.
+///
+/// Note: Assumes proper GPIO configuration, does NOT configure GPIO pins.
+///
+/// Features of the peripheral that are explicitly NOT supported by this API are:
+/// - Target mode
+/// - Interrupt Driven/Asynchronous writes/reads
+/// - DMA based writes/reads
 pub const SPI = enum(u1) {
     _,
 
     fn get_regs(spi: SPI) *volatile SpiRegs {
         return switch (@intFromEnum(spi)) {
-            0 => SPI0,
-            1 => SPI1,
+            0 => SPI0_reg,
+            1 => SPI1_reg,
         };
     }
 
-    pub fn apply(spi: SPI, comptime config: Config) void {
+    /// Initializes the SPI HW block per the Config provided
+    /// Per the API limitations discussed above, the following settings are fixed and not configurable:
+    /// - Controller Mode only
+    /// - DREQ signalling is always enabled, harmless if DMA isn't configured to listen for this
+    pub fn apply(spi: SPI, comptime config: Config) ConfigError!void {
+        if (config.data_width < 4) return ConfigError.InvalidDataWidth;
+
```

Comment:
Add a check for upper range as well, or we can replace `data_width` with an enum that exposes all legal values